File: day-02/workspace/java-oop/src/main/java/com/targetindia/model/Product.java
package com.targetindia.model;

public class Product {
    private int id;
    private String name;
    private double price;

    // best practice
    public Product() {
    }

    // overloaded / parameterized constructor
    public Product(int id, String name, double price) {
        this.id = id;
        this.name = name;
        this.price = price;
    }

    public Product(int id, String name) {
        this.id = id;
        this.name = name;
    }

    public void setId(int id) {
        if (id < 0) {
            throw new RuntimeException("Invalid value for id. Must be >= 0.");
        }
        this.id = id;
    }

    public void setName(String name) {
        if (name != null) {
            name = name.trim(); // "    asdf  "  --> "asdf" or "     " --> ""
            if (name.length() < 3 || name.length() > 50) {
                throw new RuntimeException("Length of name must be between 3 and 50 letters.");
            }
        }
        this.name = name;
    }

    public void setPrice(double price) {
        if (price < 0) {
            throw new RuntimeException("Invalid value for price. Must be >= 0.");
        }
        this.price = price;
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name.toUpperCase();
    }

    public double getPrice() {
        return price;
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", price=" + price +
                '}';
    }
}

File: day-12/workspace/kotlin-oop/src/main/kotlin/com/targetindia/model/Product.kt
package com.targetindia.model

import java.lang.RuntimeException
import java.util.*

class Product {
    var id: Int = 0
        set(value) {
            if (value < 0) {
                throw RuntimeException("id must be >=0")
            }
            field = value
        } // this is the overwritten setter --> public final void setId(int id){...}
    var name: String = ""
        set(value) {
            var len: Int = value.trim().length
            if (len < 3 || len > 25) {
                throw RuntimeException("length of name must be between 3 and 25")
            }
            field = value.trim()
        }
        get() {
            return field.uppercase(Locale.getDefault())
        }
    var price: Double = 0.0
        set(value) {
            if (value < 0.0) {
                throw RuntimeException("price must be >= 0.0")
            }
            field = value
        }


    constructor(){
    }

    constructor(id: Int, name: String, price: Double) {
        // calls to the corresponding setter methods
        this.id = id
        this.name = name
        this.price = price
    }

    constructor(id: Int, name: String) {
        this.id = id
        this.name = name
    }


    override fun toString(): String {
        return "Product(id=$id, name='$name', price=$price)"
    }

}